### DISCUSSION TITLE
Zone could not be found?

### DISCUSSION THREAD

**Discussion by Avamander** [go-acme/lego#1612] [Q&A]
I'm seeing an error saying `cloudflare: failed to find zone subsub.sub.domain.com.: Zone could not be found`. From my perspective, the zones exist and subsub.sub.domain.com resolves.

What does it very literally mean `Zone could not be found`? What is it checking? Is there a way to disable this check?

**Comment by ldez** [comment]
Since you liked my answer in the Traefik forum, I guess my answer in the forum is the answer to your question.

https://community.traefik.io/t/cloudflare-failed-to-find-zone-com-zone-could-not-be-found/8512/8?u=ldez

**Reply by Avamander** [reply]
That thread was a pointer to this repository and one of the search results.

I have been unable to determine what does it think it is "not finding" as it makes two DNS queries:
```
subsub.sub.domain.com IN SOA
_acme-challenge.subsub.sub.domain.com IN SOA
```

Both getting a NOERROR response.

**Reply by ldez** [reply]
what do you have in the `ANSWER SECTION`?

**Reply by Avamander** [reply]
For `subsub.sub.domain.com`:
```
;; ANSWER SECTION:
subsub.sub.domain.com. 259200 IN      SOA     dns.domain.com. hostmaster.domain.com. 21074661 28800 3600 2419200 86400
```

For `_acme-challenge.subsub.sub.domain.com` there is no answer section (only authority section) as it is not a standalone zone.

**Reply by ldez** [reply]
Are you using custom resolvers?
Have you tried to set resolvers to your Traefik configuration?

https://doc.traefik.io/traefik/v2.6/https/acme/#resolvers

**Reply by Avamander** [reply]
> Are you using custom resolvers?

Yes, a firewall's local recursive resolver.

> Have you tried to set resolvers to your Traefik configuration?

Third party ones? That would not work in the current use-case.

**Reply by ldez** [reply]
When you use `dig`/`drill`/`nslookup`, the command uses your custom resolvers.
You have to do the same for Traefik or override them by setting "standard" resolvers.

**Reply by Avamander** [reply]
I did try specifying the local resolver, just not any third party ones. Saw the same lookups.

**Reply by ldez** [reply]
so follow the other possibility and set by example `1.1.1.1`

**Reply by ldez** [reply]
Are we agree that you are defining the `resolvers` option in the Traefik configuration?
https://doc.traefik.io/traefik/v2.6/https/acme/#resolvers

**Reply by Avamander** [reply]
> so follow the other possibility and set by example `1.1.1.1`

It will get intercepted

> Are we agree that you are setting the resolvers option in the Traefik configuration?

Yes

**Reply by ldez** [reply]
So you have to allow Traefik to get the right information, because currently, if I follow your information, it only has errors as DNS calls answer.

You have to allow those calls or be able to give to Traefik a DNS/resolvers to get the right information.

**Reply by Avamander** [reply]
> only has errors as DNS calls answer.

The logs and tests do not agree with that though.

It seems to be expecting something different, clearly.

**Reply by ldez** [reply]
can you provide full information (logs, configuration, etc.)?

I'm sorry but you are the only person that knows your network, I cannot do more than provide explanations based on the information that you give to me.

**Reply by ldez** [reply]
If lego is not able to find a zone it's because all the recursive requests received a DNS error or an empty answer section, this is what the code does, there are no other possibilities.

So the problem comes with your firewall, I don't know your firewall and how you can configure it, maybe you can allow calls, maybe you have a local DNS in your firewall that is able to provide the right answer, or maybe something else.

**Reply by Avamander** [reply]
> can you provide full information (logs, configuration, etc.)?

The two relevant queries made shortly before the erroring according to BIND9 are just:
```
named[2215037]: client @0x7f21a4021d70 192.168.0.217#45102 (_acme-challenge.subsub.sub.domain.com): query: _acme-challenge.subsub.sub.domain.com IN SOA +E(0) (192.168.0.1)
named[2215037]: client @0x7f218cb46450 192.168.0.217#40115 (subsub.sub.domain.com): query: subsub.sub.domain.com IN SOA +E(0) (192.168.0.1)
```
While `dig`/`nslookup` provides `NOERROR` with responses as written above.

I'm not entirely sure what other details would help.

**Reply by Avamander** [reply]
> If lego is not able to find a zone it's because all the recursive requests received a DNS error or an empty answer section, this is what the code does, there are no other possibilities.

Hmm, but shouldn't the answer section be empty when, well, there's no zone for `_acme-challenge`? That seems to be how other resolvers respond as well.

**Reply by ldez** [reply]
You said:

> For `_acme-challenge.subsub.sub.domain.com` there is no answer section (only authority section) as it is not a standalone zone.

**Reply by ldez** [reply]
The answer for the zone must be like that:
```console
$ drill example.com. SOA        
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 13818
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; example.com. IN      SOA

;; ANSWER SECTION:
example.com.    429     IN      SOA     ns.icann.org. noc.dns.icann.org. 2022031411 7200 3600 1209600 3600

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 11 msec
;; SERVER: 2a02:842b:5da:c101::1
;; WHEN: Sat Mar 26 19:24:27 2022
;; MSG SIZE  rcvd: 85
```

`_acme-challenge.subsub.sub.domain.com` is not a zone so the answer section is empty.

**Reply by Avamander** [reply]
> You said:

Exactly, that is how it's supposed to be? In that aspect both internal and external 3rd party resolvers only provide an authority section (and no answer). Once a record or multiple appear under `_acme-challenge` the reply might become different, but it doesn't get that far.

**Reply by Avamander** [reply]
Once I ask for the `SOA` of the domain listed in the `AUTHORITY SECTION` a `RESPONSE SECTION` is provided, resolvers behave the same.

**Reply by ldez** [reply]
> that is how it's supposed to be?

yes, because it's not a zone, if the answer is invalid (error or empty) lego will try the next level, it's recursive.

_acme-challenge.subsub.sub.domain.com > subsub.sub.domain.com > sub.domain.com > domain.com

If all those calls fail (error or empty), you have the message: "Zone could not be found".

Remember that lego/Traefik may not have the same information as dig about resolvers.

Lego doesn't care about authority section.

**Reply by Avamander** [reply]
> Lego doesn't care about authority section.

Maybe that's the issue?

From the local resolver `dig _acme-challenge.subsub.sub.domain.com SOA` says `subsub.sub.domain.com.` as the authority, from an external perspective it's `domain.com.`. The local resolver does provide an answer for `dig subsub.sub.domain.com SOA` but the external one does not (only an authority section).

So Lego can't handle a subdomain as the authority?

**Reply by ldez** [reply]
> Maybe that's the issue?

no, because lego do SOA calls recursively:

_acme-challenge.subsub.sub.domain.com -> subsub.sub.domain.com -> sub.domain.com -> domain.com

If all those calls fail (error or empty), you have the message: "Zone could not be found".
If one call has an answer section, lego uses the domain as the zone.

**Reply by ldez** [reply]
> So Lego can't handle a subdomain as the authority?

you don't understand, I'm sorry, but I will repeat the same things once again.

lego do:
1. an SOA call to `_acme-challenge.subsub.sub.domain.com`. If error or an empty answer go to the next step, if no error/empty answer lego will use this domain as the zone.
2. an SOA call to `subsub.sub.domain.com`. If error or an empty answer go to the next step, if no error/empty answer lego will use this domain as the zone.
3. an SOA call to `sub.domain.com`. If error or an empty answer go to the next step, if no error/empty answer lego will use this domain as the zone.
4. an SOA call to `domain.com`. If error or an empty answer go to the next step, if no error/empty answer lego will use this domain as the zone.
5. if all the previous calls fail, you will see "Zone could not be found"

**Reply by Avamander** [reply]
> If all those calls fail (error or empty), you have the message: "Zone could not be found".

Well they aren't failing. Logs say, and while doing identical lookups, these queries within the entire "chain" provide a non-error response.

> an SOA call to _acme-challenge.subsub.sub.domain.com. An empty answer, go to the next step.
> an SOA call to subsub.sub.domain.com. Answer, lego will use this domain as the zone. (logs confirm this is the last query)

Only reasonable explanation, to me, is that it can't properly use a subdomain as the zone.

**Reply by ldez** [reply]
> Only reasonable explanation, to me, is that it can't properly use a subdomain as the zone.

You can, there are no limitations about that in lego, it's a classical use case.

The problem is not with lego/Traefik but with your network.

`subsub.sub.domain.com` must answer to SOA call.
You can define your "custom" resolvers in Traefik if your local DNS can give the right answer.

**Reply by Avamander** [reply]
> The problem is not with lego/Traefik but with your network.

I'd currently side with the rest of the software that seems to work fine.

> subsub.sub.domain.com must answer to SOA call.

The DNS servers in the SOA records (from internal and external) do resolve that zone. `dig subsub.sub.domain.com SOA` provides an answer.

**Reply by ldez** [reply]
> `dig subsub.sub.domain.com SOA` provides an answer.

So **Traefik/lego** has a non-empty `ANSWER SECTION`?
If yes it must work, if not it will fail.

The answer must have a non-empty `ANSWER SECTION`, other information from the response are not in the scope of an SOA expected answer.

If Traefik/lego has a valid answer, it will create the TXT records.
If Traefik/lego doesn't have a valid answer, this is because something (not in Traefik or lego) changes the answer.

The lego algorithm about zone detection is simple, there is no hidden/magical stuff.

**Reply by Avamander** [reply]
> So Traefik/lego has a non-empty `ANSWER SECTION`?

Yes.

> If yes it must work, if not it will fail.

Well...

**Reply by ldez** [reply]
> Well...

so Traefik/lego doesn't have a non-empty `ANSWER SECTION`.

**Reply by ldez** [reply]
If you still have "Zone could not be found" it's because the `ANSWER SECTION` is empty.
Sorry but it's the only possibility.

**Reply by Avamander** [reply]
I will make a packet capture, I doubt that dig sees things differently than Lego.

**Reply by ldez** [reply]
ok I'm sorry the message comes from Cloudflare.

**Reply by ldez** [reply]
Your domain seems to be not registered as a zone in Cloudflare.

The SOA calls work but when lego try to get the zone ID, Cloudflare answers with "Zone could not be found" because there is no zone defined in Cloudflare for this domain.

**Reply by ldez** [reply]
lego uses the official Cloudflare client to get the zone ID.
The zone ID is required to add/remove a TXT record.

Once again I'm sorry, I was confused by the error message.

**Reply by Avamander** [reply]
Well yes, CF only allows using domain.com and not anything below that.

**Reply by ldez** [reply]
So we have the "answer" :smile: 

I'm sorry but if it's a Cloudflare limitation, I will not be able to bypass that, the ACME challenge requires creating TXT records in a valid zone in Cloudflare.

**Reply by Avamander** [reply]
Well it could, but the record has to be placed under the primary domain (and zone).

**Reply by Avamander** [reply]
@ldez Thanks for the help with coming to this conclusion. But I think this is a bug in the process of using the SOA results for CF API. As CF API always wants you to use the "root" zone basically and all domains should be converted to that, having a local authority due to split-horizon should not change that part.

**Reply by Avamander** [reply]
@asbachb 

You can get one for a subdomain (`sub.example.com`) if your local resolver points at the *exact* CF zone name as authoritative (SOA `example.com`). If you've got a split-horizon setup and you want to request a certificate for a local domain (`subsub.sub.example.com`) where the local resolver is returning a subdomain SOA (`sub.example.com`), it will fail.

Lego makes the erroneous assumption that the value returned in SOA matches the zone name at Cloudflare, when that might not be the case.
